### ISSUE TITLE
IOException

### ISSUE THREAD

**Issue by jredfox** [offbynull/portmapper#39]
OS:Windows 10 latest
Exception: https://pastebin.com/wetU2nCi
Code: example from github

It sounds like your not getting the ipv4 dynamically yes there is a way to do this in java it's called
   public static void main(String[] args) {
        try {
            InetAddress i = InetAddress.getLocalHost();
            System.out.println(i);                  // host name and IP address
            System.out.println(i.getHostName());    // name
            System.out.println(i.getHostAddress()); // IP address only
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

**Comment by offbynull** [comment]
This error is expected on Windows. It shouldn't hinder any of the functionality provided by the library.

**Comment by jredfox** [comment]
no it's not expected on windows make it more dynamic to get the ip like internet adress.getHost

**Comment by offbynull** [comment]
The commands are run to get the IP address of the gateway. Java doesn't
provide you with a way to get gateway addresses -- that's why those
commands get run.

It's discussed at
https://github.com/offbynull/portmapper#how-does-this-library-discover-nat-pmp-and-pcp-gateway-devices

**Comment by offbynull** [comment]
PCP, NAT-PMP, and UPnP-IGD are different protocols that essentially do this
same thing: open up ports to the outside world.

UPnP-IGD is the oldest and most supported, but it's also the most
heavy/complicated. NAT-PMP is newer but has since been deprecated in favour
of PCP.

Your router may support any of the 3 protocols.

**Comment by jredfox** [comment]
Actually it does look at the source of WaifUPnP which is what I am using now.

**Comment by offbynull** [comment]
No it doesn't. UPnP does broadcasting to probe for compatible
routers/devices. NAT-PMP and PCP require a gateway address, which is
usually supplied by the OS.

The UPnP implementation in portmapper uses broadcasting to locate your
router.
The NAT-PMP/PCP implementation in portmapper uses OS commands to locate
your router/gateway.

Last I checked, there was a PCP RFC that provided some form of probing to
find the gateway, but it was far from finalized and even when it is
finalized there'll be a broad range of existing devices that won't support
it. I spent a good chunk of time struggling with this problem when I first
implemented portmapper. Unfortunately, there is no elegant solution here.
The OS commands are required if you want to use NAT-PMP/PCP. They aren't
required if you want to use UPnP.
